server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -472,7 +472,11 @@ private static void parseObjectDynamic(DocumentParserContext context, String cur
                 dynamicObjectMapper = new NoOpObjectMapper(currentFieldName, context.path().pathAsText(currentFieldName));
             } else {
                 dynamicObjectMapper = DynamicFieldsBuilder.createDynamicObjectMapper(context, currentFieldName);
-                context.addDynamicMapper(dynamicObjectMapper);
+                if (context.addDynamicMapper(dynamicObjectMapper) == false) {
+                    failIfMatchesRoutingPath(context, currentFieldName);
+                    context.parser().skipChildren();
+                    return;
+                }
```

Comment:
Sure, one place is in [parseObjectDynamic()](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java?plain=1#L475-L489), as you can se we go through the different `Dynamic` modes.
Similar code is also present in [parseArray()](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java?plain=1#L547-L565). 

The code I highlighted is in the else-else block I was referring to, IMO here for `Dynamic.TRUE` we should just add the mapper, without checks and just for `Dynamic.UNTIL_LIMIT` we should perform the check.

Reply:
I've resolved a merge conflict for this segment. Please have another look if that seems alright.

Reply:
Looks alright!